### ISSUE TITLE
nn2

### ISSUE THREAD

**Issue by juli31** [juli31/machine_learning_2017#1]

**Comment by juli31** [comment]
# coding: utf-8

# In[1]:

get_ipython().system(u'mkdir cifar10')
get_ipython().system(u'curl -o cifar-10-python.tar.gz https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz')
get_ipython().system(u'tar -xvzf cifar-10-python.tar.gz -C cifar10')


# In[1]:

get_ipython().system(u'ls')


# In[2]:

#Подключаем все нужные библиотеки

import cPickle as pickle
import numpy as np
import os
import matplotlib.pyplot as plt
import theano
import theano.tensor as T
import lasagne
from theano import tensor as T
from lasagne.nonlinearities import *
from lasagne import init

get_ipython().magic(u'matplotlib inline')
get_ipython().magic(u'pylab inline')


# In[3]:

def unpickle(file):
    import pickle
    with open(file, 'rb') as fo:
        dict = pickle.load(fo, encoding='bytes')
    return dict


# In[4]:

def load_CIFAR_batch(filename):
  """ load single batch of cifar """
  with open(filename, 'rb') as f:
    datadict = pickle.load(f)
    X = datadict['data']
    Y = datadict['labels']
    X = X.reshape(10000, 3, 32, 32).astype("float")
    Y = np.array(Y)
    return X, Y

def load_CIFAR10(ROOT):
  """ load all of cifar """
  xs = []
  ys = []
  for b in range(2,4):
    print(b)
    f = os.path.join(ROOT, 'data_batch_%d' % (b, ))
    X, Y = load_CIFAR_batch(f)
    xs.append(X)
    ys.append(Y)    
  Xtr = np.concatenate(xs)
  Ytr = np.concatenate(ys)
  del X, Y
  Xte, Yte = load_CIFAR_batch(os.path.join(ROOT, 'test_batch'))
  return Xtr, Ytr, Xte, Yte


# In[5]:

get_ipython().system(u'ls -alh cifar10')
get_ipython().system(u'ls -alh cifar10/cifar-10-batches-py')


# In[6]:

# with open("cifar10/cifar-10-batches-py/data_batch_3", "rb") as f:
#     lines = f.readlines()


# In[7]:

plt.rcParams['figure.figsize'] = (10.0, 8.0) 
cifar10_dir = './cifar10/cifar-10-batches-py'
X_train, y_train, X_test, y_test = load_CIFAR10(cifar10_dir)


# In[8]:

classes = ['plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck']
num_classes = len(classes)
samples_per_class = 7
for y, cls in enumerate(classes):
    idxs = np.flatnonzero(y_train == y)
    idxs = np.random.choice(idxs, samples_per_class, replace=False)
    for i, idx in enumerate(idxs):
        plt_idx = i * num_classes + y + 1
        plt.subplot(samples_per_class, num_classes, plt_idx)
        plt.imshow(X_train[idx].astype('uint8').transpose(1, 2, 0))
        plt.axis('off')
        if i == 0:
            plt.title(cls)
plt.show()


# In[41]:

input_X = T.tensor4("X")
target_y = T.vector("target Y integer",dtype='int32')


# In[42]:

def build_cnn(input_X=None, cnf = 64, cfs = 3):
    net = lasagne.layers.InputLayer(shape=(None, 3, 32, 32), input_var=input_X)
    net = lasagne.layers.Conv2DLayer(net, 3*cnf, filter_size = (cfs, cfs),nonlinearity = lasagne.nonlinearities.rectify, W = lasagne.init.GlorotUniform())
    net = lasagne.layers.MaxPool2DLayer(net, pool_size=(2, 2))
    net = lasagne.layers.Conv2DLayer(net, num_filters = 2*cnf, filter_size = (cfs, cfs), nonlinearity = lasagne.nonlinearities.rectify, W = lasagne.init.GlorotUniform(gain='relu'))
    net = lasagne.layers.Conv2DLayer(net, num_filters = cnf,filter_size = (cfs, cfs),nonlinearity = lasagne.nonlinearities.rectify,W = lasagne.init.GlorotUniform(gain='relu'))
    net = lasagne.layers.MaxPool2DLayer(net, pool_size=(2, 2))
    
    net = lasagne.layers.DenseLayer(
            lasagne.layers.dropout(net, p=.5),
            num_units=128,
            nonlinearity=lasagne.nonlinearities.rectify)

    net = lasagne.layers.DenseLayer(
            lasagne.layers.dropout(net, p=.5),
            num_units=10,
            nonlinearity=lasagne.nonlinearities.softmax)

    return net   
    
#    net = lasagne.layers.InputLayer(shape=(None, 3, 32, 32), input_var=input_X)
#    net = lasagne.layers.Conv2DLayer(net, 15, 3, pad='valid', W=init.Constant()) # сверточный слой
#    net = lasagne.layers.MaxPool2DLayer(net, pool_size=(2, 2))
#    net = lasagne.layers.Conv2DLayer(net, 10, 3, pad='full', W=init.Constant())  # сверточный слой
#    net = lasagne.layers.DenseLayer(net, num_units=500) # полносвязный слой
#    net = lasagne.layers.DropoutLayer(net, 0.5)         # регуляризатор
#    net = lasagne.layers.DenseLayer(net, num_units=200) # полносвязный слой
#    net = lasagne.layers.DenseLayer(net, num_units=10, nonlinearity=lasagne.nonlinearities.softmax)  # полносвязный слой


# In[43]:

def build_nn(input_X=None, cnf = 64, cfs = 3):
    net = lasagne.layers.InputLayer(shape=(None, 3, 32, 32), input_var=input_X)
    
    net = lasagne.layers.Conv2DLayer(
        net, 
        num_filters=4,
        filter_size=(6, 6),
        stride=(2, 2),
        pad=3,
        nonlinearity=lasagne.nonlinearities.rectify,
        W=lasagne.init.HeUniform())
    
    net = lasagne.layers.DenseLayer(
            net,
            num_units=40,
            nonlinearity=lasagne.nonlinearities.sigmoid,
            W=init.HeNormal())

    net = lasagne.layers.DenseLayer(
            net,
            num_units=10,
            nonlinearity=lasagne.nonlinearities.softmax,
            W=init.HeNormal())

    return net


# In[44]:

net = build_nn(input_X)

y_predicted = lasagne.layers.get_output(net, deterministic=True)
all_weights = lasagne.layers.get_all_params(net)
print all_weights


# In[45]:

#функция ошибки и точности будет прямо внутри
loss = lasagne.objectives.categorical_crossentropy(y_predicted, target_y)
loss = loss.mean()
accuracy = lasagne.objectives.categorical_accuracy(y_predicted, target_y)
accuracy = accuracy.mean()


# In[46]:

updates = lasagne.updates.adam(loss, all_weights, learning_rate=0.01)


# In[47]:

#функция, делает updates и возвращащет значение функции потерь и точности
train_fun = theano.function([input_X, target_y], [loss, accuracy], updates = updates, allow_input_downcast=True)
test_fun = theano.function([input_X, target_y], accuracy, allow_input_downcast=True) # точность без обновления весов, для теста


# In[48]:

def iterate_minibatches(inputs, targets, batchsize, shuffle=False):
    assert len(inputs) == len(targets)
    if shuffle:
        indices = np.arange(len(inputs))
        np.random.shuffle(indices)
    for start_idx in range(0, len(inputs) - batchsize + 1, batchsize):
        if shuffle:
            excerpt = indices[start_idx:start_idx + batchsize]
        else:
            excerpt = slice(start_idx, start_idx + batchsize)
        yield inputs[excerpt], targets[excerpt]


# In[49]:

import time

num_epochs = 10 #количество проходов по данным

batch_size = 50 #размер мини-батча

for epoch in range(num_epochs):
    # In each epoch, we do a full pass over the training data:
    train_err = 0
    train_acc = 0
    train_batches = 0
    start_time = time.time()
    for batch in iterate_minibatches(X_train, y_train,batch_size):
        inputs, targets = batch
        train_res = train_fun(inputs, targets)
        train_err_batch, train_acc_batch = train_res
        train_err += train_err_batch
        train_acc += train_acc_batch
        train_batches += 1

    # And a full pass over the validation data:
    val_acc = 0
    val_batches = 0
    for batch in iterate_minibatches(X_test, y_test, batch_size):
        inputs, targets = batch
        val_acc += test_fun(inputs, targets)
        val_batches += 1

    # Then we print the results for this epoch:
    print("Epoch {} of {} took {:.3f}s".format(epoch + 1, num_epochs, time.time() - start_time))
    print("  training loss (in-iteration):\t\t{:.6f}".format(train_err / train_batches))
    print("  train accuracy:\t\t{:.2f} %".format(train_acc / train_batches * 100))
    print("  validation accuracy:\t\t{:.2f} %".format(val_acc / val_batches * 100))


# In[23]:

test_acc = 0
test_batches = 0
for batch in iterate_minibatches(X_test, y_test, 500):
    inputs, targets = batch
    acc = test_fun(inputs, targets)
    test_acc += acc
    test_batches += 1
print("Final results:")
print("  test accuracy:\t\t{:.2f} %".format(
    test_acc / test_batches * 100))

if test_acc / test_batches * 100 > 80:
    print "Achievement unlocked: колдун 80 уровня"
else:
    print "Нужно больше магии!"


# In[ ]:




# In[ ]:




# In[ ]:




# In[ ]:




# In[ ]:
